<a href="https://colab.research.google.com/github/blazaropinto/portfolio_PDA_Data_Science/blob/development/Preparing_data_for_analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Removing duplicate records from a data set and replacing null values where necessary

### Data set 1 - Calorie burning


In [14]:
import pandas as pd
url1_raw = 'https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/Data%20Cleaning%20Data%20Sets/dirtydata.csv'
dataset1 = pd.read_csv(url1_raw)
display(dataset1.head(), dataset1.shape)

Duration          Date  Pulse  Maxpulse  Calories
0        60  '2020/12/01'    110       130     409.1
1        60  '2020/12/02'    117       145     479.0
2        60  '2020/12/03'    103       135     340.0
3        45  '2020/12/04'    109       175     282.4
4        45  '2020/12/05'    117       148     406.0

(32, 5)

* Start by inspecting the data a bit further

In [15]:
dataset1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32 entries, 0 to 31
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Duration  32 non-null     int64  
 1   Date      31 non-null     object 
 2   Pulse     32 non-null     int64  
 3   Maxpulse  32 non-null     int64  
 4   Calories  30 non-null     float64
dtypes: float64(1), int64(3), object(1)
memory usage: 1.4+ KB


* Make sure there are no duplicated rows (considering all columns), by deleteting them if any

In [16]:
dataset1.drop_duplicates(inplace=True)
dataset1.shape

(31, 5)

* The dtype of the column date is `object`, probably most of them are strings. We can use `.strptime()` on each element or just use the pandas function`.to_datetime()`

In [17]:
dataset1.Date = pd.to_datetime(dataset1.Date)
dataset1.dtypes

Duration             int64
Date        datetime64[ns]
Pulse                int64
Maxpulse             int64
Calories           float64
dtype: object

* There are only 3 missing values, take a look at these rows

In [19]:
dataset1[dataset1.isna().any(axis=1)]

Duration       Date  Pulse  Maxpulse  Calories
18        45 2020-12-18     90       112       NaN
22        45        NaT    100       119     282.0
28        60 2020-12-28    103       132       NaN

* The missing value in the date column can't be filled or replaced with anything, this information seems to be key for the record to mean anything so we can drop the row
* Once this is removed  fill the missing values with an appropriate statistic (use `.describe()` to inspect and decide)

In [20]:
dataset1.dropna(subset=['Date'], inplace=True)
dataset1.describe()

Duration       Pulse    Maxpulse    Calories
count   30.000000   30.000000   30.000000   28.000000
mean    69.500000  103.733333  129.100000  307.417857
std     72.258516    8.042702   13.215325   67.412594
min     30.000000   90.000000  101.000000  195.100000
25%     60.000000  100.000000  120.750000  250.525000
50%     60.000000  103.000000  128.500000  300.000000
75%     60.000000  107.500000  132.750000  350.025000
max    450.000000  130.000000  175.000000  479.000000

* The mean and median seem to be fairly similar in all the numeric values

In [21]:
clean_dataset1 = dataset1.copy()
calories_mean = clean_dataset1.Calories.mean()
clean_dataset1.Calories = clean_dataset1.Calories.fillna(calories_mean)
display(clean_dataset1.info(), clean_dataset1.describe())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 30 entries, 0 to 31
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype         
---  ------    --------------  -----         
 0   Duration  30 non-null     int64         
 1   Date      30 non-null     datetime64[ns]
 2   Pulse     30 non-null     int64         
 3   Maxpulse  30 non-null     int64         
 4   Calories  30 non-null     float64       
dtypes: datetime64[ns](1), float64(1), int64(3)
memory usage: 1.4 KB


None

Duration       Pulse    Maxpulse    Calories
count   30.000000   30.000000   30.000000   30.000000
mean    69.500000  103.733333  129.100000  307.417857
std     72.258516    8.042702   13.215325   65.046498
min     30.000000   90.000000  101.000000  195.100000
25%     60.000000  100.000000  120.750000  251.350000
50%     60.000000  103.000000  128.500000  300.000000
75%     60.000000  107.500000  132.750000  343.975000
max    450.000000  130.000000  175.000000  479.000000

In [ ]:
from google.colab import files
clean_dataset1.to_csv('dataset1.csv')
files.download('dataset1.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

### Data set 2 - Presidential election voting
---



In [23]:
import pandas as pd
url2_raw = 'https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/Data%20Cleaning%20Data%20Sets/presDirty.csv'
dataset2 = pd.read_csv(url2_raw)
dataset2.head()

cycle     branch  ...         timestamp simulations
0   2020  President  ...  03/11/2020 00:00       40000
1   2020  President  ...  02/11/2020 23:45       40000
2   2020  President  ...  02/11/2020 23:45       40000
3   2020  President  ...  31/10/2020 22:23       40000
4   2020  President  ...  30/10/2020 21:34       40000

[5 rows x 16 columns]

* Start by inspecting the data

In [24]:
dataset2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 156 entries, 0 to 155
Data columns (total 16 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   cycle                       156 non-null    int64  
 1   branch                      156 non-null    object 
 2   model                       156 non-null    object 
 3   modeldate                   156 non-null    object 
 4   candidate_inc               156 non-null    object 
 5   candidate_chal              156 non-null    object 
 6   nat_voteshare_other_hi      156 non-null    float64
 7   national_voteshare_inc_lo   156 non-null    float64
 8   national_voteshare_chal_lo  156 non-null    float64
 9   national_voteshare_3rd_lo   0 non-null      float64
 10  nat_voteshare_other_lo      156 non-null    float64
 11  national_turnout            59 non-null     float64
 12  national_turnout_hi         59 non-null     float64
 13  national_turnout_lo         59 non-

* There is an empty column that can be deleted. 
* There are another 3 columns with roughly 2/3 on its values missing. Take a look at the statistics.

In [25]:
dataset2.drop(columns=['national_voteshare_3rd_lo'], inplace=True)
dataset2.describe()

cycle  nat_voteshare_other_hi  ...  national_turnout_lo  simulations
count   156.0              156.000000  ...         5.900000e+01        156.0
mean   2020.0                2.004223  ...         1.343265e+08      40000.0
std       0.0                0.061918  ...         5.684473e+06          0.0
min    2020.0                1.835479  ...         1.310000e+08      40000.0
25%    2020.0                1.974398  ...         1.310000e+08      40000.0
50%    2020.0                2.030170  ...         1.310000e+08      40000.0
75%    2020.0                2.048704  ...         1.330000e+08      40000.0
max    2020.0                2.085074  ...         1.475607e+08      40000.0

[8 rows x 9 columns]

--> in the 3 columns the median value is equal to its minimum value, but the mean is higher than the upper quartile. Inspect more the dataset before making a decision.

* Many of this columns don't seem to have the most efficient datatype. Look at the number of different values on each columns to see if they should be treated as categories

In [26]:
for column in dataset2.columns:
  print(column, ': ',
        len(dataset2[column].unique()),
        'different values')

cycle :  1 different values
branch :  1 different values
model :  1 different values
modeldate :  155 different values
candidate_inc :  2 different values
candidate_chal :  1 different values
nat_voteshare_other_hi :  154 different values
national_voteshare_inc_lo :  155 different values
national_voteshare_chal_lo :  155 different values
nat_voteshare_other_lo :  154 different values
national_turnout :  8 different values
national_turnout_hi :  8 different values
national_turnout_lo :  8 different values
timestamp :  136 different values
simulations :  1 different values


* All the columns with just 1 value could also be deleted as they don't provide valuable information, but keep the original dataset separated in case it's needed.
* look at the 2 values of 'candidate_inc' 


In [27]:
clean_dataset2 = dataset2.copy()
clean_dataset2.drop(columns=['cycle', 'branch', 'model', 'candidate_chal', 'simulations'], inplace=True)

In [28]:
clean_dataset2.candidate_inc.unique()

array(['Trump', ' '], dtype=object)

In [29]:
clean_dataset2.candidate_inc.value_counts()

Trump    154
           2
Name: candidate_inc, dtype: int64

--> so these couple of 'blanks' are actually unrecognized missing values

* convert the 'timestamp' and the 'modeldate' columns into the correct pandas type and calculate the range of time of the data

In [30]:
clean_dataset2.timestamp = pd.to_datetime(clean_dataset2.timestamp)
print('The data was collected from', clean_dataset2.timestamp.min(), 'to', clean_dataset2.timestamp.max())

The data was collected from 2020-01-09 19:06:00 to 2020-12-10 21:07:00


--> This suggest the missing data in candidate_inc is also 'Trump' and there is a unique value in the column so it can be also eliminated

In [31]:
clean_dataset2[['modeldate', 'timestamp']].head()

modeldate           timestamp
0  11/03/2020 2020-03-11 00:00:00
1  11/02/2020 2020-02-11 23:45:00
2  11/02/2020 2020-02-11 23:45:00
3  10/31/2020 2020-10-31 22:23:00
4  10/30/2020 2020-10-30 21:34:00

--> 'modeldate' seems to be the data from the timestamp in US format, so it doesn't provide any information and could be deleted

In [32]:
clean_dataset2.drop(['candidate_inc', 'modeldate'], axis=1, inplace=True)

* Get more info about the 3 columns that still have missing values

In [33]:
clean_dataset2.sort_values('timestamp', inplace=True)

In [38]:
clean_dataset2.groupby('national_turnout').median()

nat_voteshare_other_hi  ...  national_turnout_lo
national_turnout                          ...                     
141000000.0                     1.976343  ...          131000000.0
143000000.0                     1.903847  ...          133000000.0
146000000.0                     1.900200  ...          136000000.0
154000000.0                     1.882668  ...          143000000.0
156723120.0                     1.847821  ...          146239008.0
157972048.0                     1.842986  ...          147487920.0
158044784.0                     1.835479  ...          147560672.0

[7 rows x 6 columns]

In [39]:
clean_dataset2.groupby('national_turnout_hi').median()

nat_voteshare_other_hi  ...  national_turnout_lo
national_turnout_hi                          ...                     
151000000.0                        1.976343  ...          131000000.0
153000000.0                        1.903847  ...          133000000.0
156000000.0                        1.900200  ...          136000000.0
164000000.0                        1.882668  ...          143000000.0
167175712.0                        1.847821  ...          146239008.0
168424640.0                        1.842986  ...          147487920.0
168497376.0                        1.835479  ...          147560672.0

[7 rows x 6 columns]

In [40]:
clean_dataset2.groupby('national_turnout_lo').agg(['min', 'median', 'max'])

nat_voteshare_other_hi            ... national_turnout_hi             
                                       min    median  ...              median          max
national_turnout_lo                                   ...                                 
131000000.0                       1.911543  1.976343  ...         151000000.0  151000000.0
133000000.0                       1.899508  1.903847  ...         153000000.0  153000000.0
136000000.0                       1.900200  1.900200  ...         156000000.0  156000000.0
143000000.0                       1.866135  1.882668  ...         164000000.0  164000000.0
146239008.0                       1.847821  1.847821  ...         167175712.0  167175712.0
147487920.0                       1.842986  1.842986  ...         168424640.0  168424640.0
147560672.0                       1.835479  1.835479  ...         168497376.0  168497376.0

[7 rows x 18 columns]

In [41]:
clean_dataset2[clean_dataset2.isna().any(axis=1)]

nat_voteshare_other_hi  ...           timestamp
63                 2.036718  ... 2020-01-09 19:06:00
62                 2.043273  ... 2020-02-09 21:52:00
61                 2.024914  ... 2020-03-09 21:55:00
60                 2.035542  ... 2020-04-09 22:34:00
59                 2.040212  ... 2020-05-09 22:17:00
..                      ...  ...                 ...
148                2.051627  ... 2020-10-08 11:59:00
152                2.039343  ... 2020-10-08 12:03:00
85                 2.034930  ... 2020-10-08 23:50:00
84                 2.028735  ... 2020-11-08 19:08:00
83                 2.027544  ... 2020-12-08 23:41:00

[97 rows x 8 columns]

Eliminating 97 rows from a 156-row dataframe is not an option. We've seen there are only 7 unique values on each of these 3 columns, and they all seem to be quite correlated with the columns 'nat_voteshare_other_hi' and 'national_voteshare_inc_lo'. 

* One option would be deleting these 3 columns if these information is not going to be used.

* Another is replacing the missing values with the most suitable ones. To keep it simple for this exercise this will be sorted by splitting the dataframe in two and through replacement of the missing values on each columns of each dataset portion with the most suitable value (taken fron the statistics above analized with `.describe()`).
    * par1 of the dataframe will contain rows with lower values in the column 'nat_voteshare_other_hi' and missing values in the 3 turnout columns will be higher (the upper percentile)
    * par2 of the dataframe will contain rows with higher values in the column 'nat_voteshare_other_hi' and missing values in the 3 turnout columns will be lower (the median)

* instead of using the keyword `inplace` in `fillna()` in this case it works better assigning the changed dataset portions to new variables to append them afterwards

In [42]:
clean_dataset2_part1 = clean_dataset2[clean_dataset2['nat_voteshare_other_hi'] < clean_dataset2['nat_voteshare_other_hi'].quantile(0.6)]
clean_dataset2_part1 = clean_dataset2_part1.fillna(value= {'national_turnout': 1.430000e+08, 'national_turnout_hi': 1.530000e+08, 'national_turnout_lo': 1.330000e+08})

clean_dataset2_part2 = clean_dataset2[clean_dataset2['nat_voteshare_other_hi'] >= clean_dataset2['nat_voteshare_other_hi'].quantile(0.6)]
clean_dataset2_part2 = clean_dataset2_part2.fillna(value= {'national_turnout': 1.410000e+08	, 'national_turnout_hi': 1.510000e+08	, 'national_turnout_lo': 1.310000e+08})

clean_dataset_new = clean_dataset2_part1.append([clean_dataset2_part2])
clean_dataset_new.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 156 entries, 63 to 148
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype         
---  ------                      --------------  -----         
 0   nat_voteshare_other_hi      156 non-null    float64       
 1   national_voteshare_inc_lo   156 non-null    float64       
 2   national_voteshare_chal_lo  156 non-null    float64       
 3   nat_voteshare_other_lo      156 non-null    float64       
 4   national_turnout            156 non-null    float64       
 5   national_turnout_hi         156 non-null    float64       
 6   national_turnout_lo         156 non-null    float64       
 7   timestamp                   156 non-null    datetime64[ns]
dtypes: datetime64[ns](1), float64(7)
memory usage: 11.0 KB


In [43]:
clean_dataset_new.describe()

nat_voteshare_other_hi  ...  national_turnout_lo
count              156.000000  ...         1.560000e+02
mean                 2.004223  ...         1.326940e+08
std                  0.061918  ...         3.780575e+06
min                  1.835479  ...         1.310000e+08
25%                  1.974398  ...         1.310000e+08
50%                  2.030170  ...         1.310000e+08
75%                  2.048704  ...         1.330000e+08
max                  2.085074  ...         1.475607e+08

[8 rows x 7 columns]

--> We can see here that the statistic summary looks similar to the summary of the dataset before replacing the missig values (after a few attempts).

In [44]:
from google.colab import files
clean_dataset2.to_csv('dataset2.csv') 
files.download('dataset2.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

### Sorting and cleaning
---

##### Check for null values 
---

1. Read data from the file housing_in_london_yearly_variables.csv from this link: https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv 
2. check if any NA values exist in the dataframe and print the result 
3. use df.info() to see which columns have null entries (*Hint: if the non-null count is less than total entries, column contains missing/NA entries*)  

**Test output**:
True
.info shows median_salary, life_satisfaction, recycling_pct, population_size, number_of_jobs, area_size, no_of_houses all less than total rows (1071) 



In [45]:
url_london = 'https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv'

df_london = pd.read_csv(url_london)
df_london.head()

code                  area  ... no_of_houses  borough_flag
0  E09000001        city of london  ...          NaN             1
1  E09000002  barking and dagenham  ...          NaN             1
2  E09000003                barnet  ...          NaN             1
3  E09000004                bexley  ...          NaN             1
4  E09000005                 brent  ...          NaN             1

[5 rows x 12 columns]

In [46]:
df_london.isnull().values.any()

True

In [47]:
df_london.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1071 entries, 0 to 1070
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   code               1071 non-null   object 
 1   area               1071 non-null   object 
 2   date               1071 non-null   object 
 3   median_salary      1049 non-null   float64
 4   life_satisfaction  352 non-null    float64
 5   mean_salary        1071 non-null   object 
 6   recycling_pct      860 non-null    object 
 7   population_size    1018 non-null   float64
 8   number_of_jobs     931 non-null    float64
 9   area_size          666 non-null    float64
 10  no_of_houses       666 non-null    float64
 11  borough_flag       1071 non-null   int64  
dtypes: float64(6), int64(1), object(5)
memory usage: 100.5+ KB


* `.values` converts the df into a np array

In [48]:
df_london.isnull().values

array([[False, False, False, ...,  True,  True, False],
       [False, False, False, ...,  True,  True, False],
       [False, False, False, ...,  True,  True, False],
       ...,
       [False, False, False, ...,  True,  True, False],
       [False, False, False, ...,  True,  True, False],
       [False, False, False, ...,  True,  True, False]])

* `.any()` tells if there is any `True` in each column of the df

In [49]:
df_london.isnull().any()

code                 False
area                 False
date                 False
median_salary         True
life_satisfaction     True
mean_salary          False
recycling_pct         True
population_size       True
number_of_jobs        True
area_size             True
no_of_houses          True
borough_flag         False
dtype: bool

In [50]:
df_london.isnull().any().any() # same output as .any() used once on an array

True

* both fuctions 'isnull()` and `isna()' are the same (https://pandas-docs.github.io/pandas-docs-travis/reference/api/pandas.DataFrame.isnull.html)

In [51]:
df_london.isna().any()  

code                 False
area                 False
date                 False
median_salary         True
life_satisfaction     True
mean_salary          False
recycling_pct         True
population_size       True
number_of_jobs        True
area_size             True
no_of_houses          True
borough_flag         False
dtype: bool

In [52]:
df_london.isna().sum()  

code                   0
area                   0
date                   0
median_salary         22
life_satisfaction    719
mean_salary            0
recycling_pct        211
population_size       53
number_of_jobs       140
area_size            405
no_of_houses         405
borough_flag           0
dtype: int64

##### Remove null values 
---

1. Remove rows with NA values for `life_satisfaction` (use [ ] even if only one column in list)
2. Remove all NA values across whole dataframe 

**Test output**:  
1.  Row count reduced to 352 rows
2.  Row count reduced to 267 rows

* Remove rows with NA values for `life_satisfaction` (use [ ] even if only one column in list)

In [ ]:
part1 = df_london.dropna(subset=['life_satisfaction'])
print('Remaining rows:', part1.shape[0])

Remaining rows: 352


* Remove all NA values across whole dataframe

In [ ]:
part2 = df_london.dropna()
print('Remaining rows:', part2.shape[0])

Remaining rows: 267


##### Removing duplicate entries 
---

remove duplicate `area` entries keeping first instance  

**Test output**:  
 Dataframe now contains 50 rows all with date 1999-12-*01* 

In [53]:
unique_entries = df_london.drop_duplicates(subset=['area'])
display(unique_entries.shape,unique_entries.date.unique()) 

(51, 12)

array(['1999-12-01'], dtype=object)

In [54]:
unique_entries.head()

code                  area  ... no_of_houses  borough_flag
0  E09000001        city of london  ...          NaN             1
1  E09000002  barking and dagenham  ...          NaN             1
2  E09000003                barnet  ...          NaN             1
3  E09000004                bexley  ...          NaN             1
4  E09000005                 brent  ...          NaN             1

[5 rows x 12 columns]

# Normalising values where necessary

# Correcting the format of dates where necessary

# Resetting data values

# Joining data from 2 different sources

# Merging and joining data